shell/src/main/java/alluxio/cli/fs/command/CpCommand.java

```diff
@@ -61,6 +69,108 @@
           .hasArg(false)
           .desc("copy files in subdirectories recursively")
           .build();
+  private static final Option THREAD_OPTION =
+      Option.builder("t")
+          .required(false)
+          .hasArg(true)
+          .numberOfArgs(1)
+          .argName("threads")
+          .type(Integer.class)
+          .desc("Number of threads used to copy files in parallel")
+          .build();
+
+  /**
+   * A thread pool for asynchronous copy.
+   *
+   * Copy tasks can send messages to an output stream in a thread safe way.
+   */
+  @ThreadSafe
+  private final class CopyThreadPool {
+    private static final String MESSAGE_DONE = "#";
+
+    private ThreadPoolExecutor mPool;
+    private BlockingQueue<String> mMessages;
+    private PrintStream mOutput;
+    private Thread mPrinter;
+    private AlluxioURI mPath;
+
+    /**
+     * Creates a new thread pool with the specified number of threads,
+     * specify the output stream for tasks to send messages to, and
+     * starts the background thread for printing messages.
+     *
+     * NOTE: needs to call {@link #shutdown()} to release resources.
+     *
+     * @param threads number of threads
+     * @param out the output stream for tasks to send messages to
+     * @param path the path to delete on shutdown when it's empty, otherwise can be {@code null}
+     */
+    public CopyThreadPool(int threads, PrintStream out, AlluxioURI path) {
+      mPool = new ThreadPoolExecutor(threads, threads,
+          1, TimeUnit.SECONDS, new ArrayBlockingQueue<>(threads * 2),
+          new ThreadPoolExecutor.CallerRunsPolicy());
+      mMessages = new LinkedBlockingQueue<>();
+      mOutput = out;
+      mPrinter = new Thread(() -> {
+        while (!Thread.currentThread().isInterrupted()) {
+          try {
+            String message = mMessages.take();
+            if (message.equals(MESSAGE_DONE)) {
+              break;
+            }
+            mOutput.println(message);
```

Comment:
is this thread safe?

Reply:
since it is in the same printer thread, it should be.